File: ios-lister/CloudStoreLister/Logic/CSVImportLogic.swift
import Foundation

enum CSVImportError: LocalizedError, Equatable {
    case emptyFile
    case missingProductIdColumn
    case noValidRows

    var errorDescription: String? {
        switch self {
        case .emptyFile:
            return "The file is empty."
        case .missingProductIdColumn:
            return "Select which column contains Product ID."
        case .noValidRows:
            return "No rows with a valid Product ID were found."
        }
    }
}

enum CSVImportableField: String, CaseIterable, Identifiable {
    case productId = "Product ID"
    case barcode = "Barcode"
    case name = "Name"
    case productType = "Type"
    case manufacturer = "Manufacturer"
    case price = "Price"
    case stock = "Stock"
    case pullCount = "Pull Count"

    var id: String { rawValue }

    static var optionalFields: [CSVImportableField] {
        allCases.filter { $0 != .productId }
    }
}

struct ParsedCSV: Equatable {
    let headers: [String]
    let rows: [[String]]
}

struct CSVFieldMapping: Equatable {
    var productIdColumn: String?
    var enabledFields: Set<CSVImportableField> = []
    var columnByField: [CSVImportableField: String] = [:]
}

enum CSVImportLogic {
    static func parseCSV(_ content: String) -> ParsedCSV? {
        let lines = nonEmptyLines(from: content)
        guard !lines.isEmpty else { return nil }

        let firstFields = parseCSVLine(lines[0])
        if looksLikeHeader(firstFields) {
            return ParsedCSV(
                headers: firstFields,
                rows: lines.dropFirst().map(parseCSVLine)
            )
        }

        let width = firstFields.count
        let headers = (0..<width).map { "Column \($0 + 1)" }
        return ParsedCSV(headers: headers, rows: lines.map(parseCSVLine))
    }

    static func parseCSVLine(_ line: String) -> [String] {
        var fields: [String] = []
        var current = ""
        var inQuotes = false
        for char in line {
            if char == "\"" {
                inQuotes.toggle()
            } else if char == "," && !inQuotes {
                fields.append(current.trimmingCharacters(in: .whitespacesAndNewlines))
                current = ""
            } else {
                current.append(char)
            }
        }
        fields.append(current.trimmingCharacters(in: .whitespacesAndNewlines))
        return fields
    }

    static func suggestedMapping(for parsed: ParsedCSV) -> CSVFieldMapping {
        var mapping = CSVFieldMapping()
        mapping.productIdColumn = guessProductIdColumn(from: parsed.headers)

        for field in CSVImportableField.optionalFields {
            if let column = guessColumn(for: field, in: parsed.headers) {
                mapping.enabledFields.insert(field)
                mapping.columnByField[field] = column
            }
        }
        return mapping
    }

    static func buildItems(parsed: ParsedCSV, mapping: CSVFieldMapping) throws -> [InventoryListItem] {
        guard let productIdColumn = mapping.productIdColumn,
              let productIdIndex = parsed.headers.firstIndex(of: productIdColumn) else {
            throw CSVImportError.missingProductIdColumn
        }

        var items: [InventoryListItem] = []
        for row in parsed.rows {
            guard productIdIndex < row.count,
                  let productId = parseProductId(row[productIdIndex]) else { continue }

            let pullCount = max(1, intValue(field: .pullCount, in: row, parsed: parsed, mapping: mapping) ?? 1)
            let stockLabel = stringValue(field: .stock, in: row, parsed: parsed, mapping: mapping) ?? "—"
            let stockEmphasis = stockLabel.localizedCaseInsensitiveContains("out")

            items.append(
                InventoryListItem(
                    productId: productId,
                    barcode: stringValue(field: .barcode, in: row, parsed: parsed, mapping: mapping),
                    name: stringValue(field: .name, in: row, parsed: parsed, mapping: mapping) ?? "",
                    productType: stringValue(field: .productType, in: row, parsed: parsed, mapping: mapping),
                    manufacturer: stringValue(field: .manufacturer, in: row, parsed: parsed, mapping: mapping),
                    priceLabel: stringValue(field: .price, in: row, parsed: parsed, mapping: mapping) ?? "—",
                    stockLabel: stockLabel,
                    stockEmphasis: stockEmphasis,
                    pullCount: pullCount
                )
            )
        }

        guard !items.isEmpty else { throw CSVImportError.noValidRows }
        return items
    }

    static func applyImport(
        items: [InventoryListItem],
        toListId: UUID,
        in lists: [InventoryNamedList]
    ) -> [InventoryNamedList] {
        var updated = lists
        for item in items {
            updated = ListStoreLogic.addItem(item, toListId: toListId, in: updated)
        }
        return updated
    }

    private static func nonEmptyLines(from content: String) -> [String] {
        content
            .components(separatedBy: .newlines)
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }
    }

    private static func looksLikeHeader(_ fields: [String]) -> Bool {
        let normalized = fields.map { normalizeHeader($0) }
        if normalized.contains(where: { $0.contains("product") && $0.contains("id") }) { return true }
        let keywords = ["name", "barcode", "type", "manufacturer", "price", "stock", "pull"]
        return normalized.contains { field in keywords.contains(where: { field.contains($0) }) }
    }

    private static func guessProductIdColumn(from headers: [String]) -> String? {
        headers.first { header in
            let normalized = normalizeHeader(header)
            return normalized == "productid"
                || normalized == "product id"
                || normalized == "id"
                || (normalized.contains("product") && normalized.contains("id"))
        }
    }

    static func guessColumn(for field: CSVImportableField, in headers: [String]) -> String? {
        let candidates: [String]
        switch field {
        case .productId:
            return guessProductIdColumn(from: headers)
        case .barcode:
            candidates = ["barcode", "upc", "ean", "sku"]
        case .name:
            candidates = ["name", "product name", "description"]
        case .productType:
            candidates = ["type", "product type", "category"]
        case .manufacturer:
            candidates = ["manufacturer", "mfr", "brand"]
        case .price:
            candidates = ["price", "regular price", "sale price"]
        case .stock:
            candidates = ["stock", "qty", "quantity", "on hand"]
        case .pullCount:
            candidates = ["pull count", "pull", "count"]
        }
        return headers.first { header in
            let normalized = normalizeHeader(header)
            return candidates.contains(normalized)
        }
    }

    private static func normalizeHeader(_ header: String) -> String {
        header
            .lowercased()
            .replacingOccurrences(of: "_", with: " ")
            .replacingOccurrences(of: "-", with: " ")
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private static func parseProductId(_ raw: String) -> Int? {
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        if let value = Int(trimmed) { return value }
        let digits = trimmed.filter(\.isNumber)
        guard !digits.isEmpty, let value = Int(digits) else { return nil }
        return value
    }

    private static func stringValue(
        field: CSVImportableField,
        in row: [String],
        parsed: ParsedCSV,
        mapping: CSVFieldMapping
    ) -> String? {
        guard mapping.enabledFields.contains(field),
              let column = mapping.columnByField[field],
              let index = parsed.headers.firstIndex(of: column),
              index < row.count else { return nil }
        let value = row[index].trimmingCharacters(in: .whitespacesAndNewlines)
        return value.isEmpty ? nil : value
    }

    private static func intValue(
        field: CSVImportableField,
        in row: [String],
        parsed: ParsedCSV,
        mapping: CSVFieldMapping
    ) -> Int? {
        guard let string = stringValue(field: field, in: row, parsed: parsed, mapping: mapping) else { return nil }
        return Int(string.filter(\.isNumber))
    }
}

File: android-lister/app/src/main/java/com/cloudstore/lister/domain/CsvImportLogic.kt
package com.cloudstore.lister.domain

import com.cloudstore.lister.data.InventoryListItem
import com.cloudstore.lister.data.InventoryNamedList

sealed class CsvImportError(message: String) : Exception(message) {
    data object EmptyFile : CsvImportError("The file is empty.")
    data object MissingProductIdColumn : CsvImportError("Select which column contains Product ID.")
    data object NoValidRows : CsvImportError("No rows with a valid Product ID were found.")
}

enum class CsvImportableField(val label: String) {
    ProductId("Product ID"),
    Barcode("Barcode"),
    Name("Name"),
    ProductType("Type"),
    Manufacturer("Manufacturer"),
    Price("Price"),
    Stock("Stock"),
    PullCount("Pull Count"),
    ;

    companion object {
        val optionalFields: List<CsvImportableField> = entries.filter { it != ProductId }
    }
}

data class ParsedCsv(
    val headers: List<String>,
    val rows: List<List<String>>,
)

data class CsvFieldMapping(
    var productIdColumn: String? = null,
    val enabledFields: MutableSet<CsvImportableField> = mutableSetOf(),
    val columnByField: MutableMap<CsvImportableField, String> = mutableMapOf(),
)

object CsvImportLogic {
    fun parseCsv(content: String): ParsedCsv? {
        val lines = nonEmptyLines(content)
        if (lines.isEmpty()) return null
        val firstFields = parseCsvLine(lines.first())
        return if (looksLikeHeader(firstFields)) {
            ParsedCsv(
                headers = firstFields,
                rows = lines.drop(1).map(::parseCsvLine),
            )
        } else {
            val width = firstFields.size
            val headers = (0 until width).map { "Column ${it + 1}" }
            ParsedCsv(headers = headers, rows = lines.map(::parseCsvLine))
        }
    }

    fun parseCsvLine(line: String): List<String> {
        val fields = mutableListOf<String>()
        val current = StringBuilder()
        var inQuotes = false
        for (char in line) {
            when {
                char == '"' -> inQuotes = !inQuotes
                char == ',' && !inQuotes -> {
                    fields.add(current.toString().trim())
                    current.clear()
                }
                else -> current.append(char)
            }
        }
        fields.add(current.toString().trim())
        return fields
    }

    fun suggestedMapping(parsed: ParsedCsv): CsvFieldMapping {
        val mapping = CsvFieldMapping(productIdColumn = guessProductIdColumn(parsed.headers))
        CsvImportableField.optionalFields.forEach { field ->
            guessColumn(field, parsed.headers)?.let { column ->
                mapping.enabledFields.add(field)
                mapping.columnByField[field] = column
            }
        }
        return mapping
    }

    fun buildItems(parsed: ParsedCsv, mapping: CsvFieldMapping): List<InventoryListItem> {
        val productIdColumn = mapping.productIdColumn
            ?: throw CsvImportError.MissingProductIdColumn
        val productIdIndex = parsed.headers.indexOf(productIdColumn)
        if (productIdIndex < 0) throw CsvImportError.MissingProductIdColumn

        val items = mutableListOf<InventoryListItem>()
        for (row in parsed.rows) {
            if (productIdIndex >= row.size) continue
            val productId = parseProductId(row[productIdIndex]) ?: continue
            val pullCount = maxOf(1, intValue(CsvImportableField.PullCount, row, parsed, mapping) ?: 1)
            val stockLabel = stringValue(CsvImportableField.Stock, row, parsed, mapping) ?: "—"
            val stockEmphasis = stockLabel.contains("out", ignoreCase = true)
            items.add(
                InventoryListItem(
                    productId = productId,
                    barcode = stringValue(CsvImportableField.Barcode, row, parsed, mapping),
                    name = stringValue(CsvImportableField.Name, row, parsed, mapping).orEmpty(),
                    productType = stringValue(CsvImportableField.ProductType, row, parsed, mapping),
                    manufacturer = stringValue(CsvImportableField.Manufacturer, row, parsed, mapping),
                    priceLabel = stringValue(CsvImportableField.Price, row, parsed, mapping) ?: "—",
                    stockLabel = stockLabel,
                    stockEmphasis = stockEmphasis,
                    pullCount = pullCount,
                ),
            )
        }
        if (items.isEmpty()) throw CsvImportError.NoValidRows
        return items
    }

    fun applyImport(
        items: List<InventoryListItem>,
        toListId: String,
        lists: List<InventoryNamedList>,
    ): List<InventoryNamedList> {
        var updated = lists
        for (item in items) {
            updated = ListStoreLogic.addItem(item, toListId, updated)
        }
        return updated
    }

    fun guessColumn(field: CsvImportableField, headers: List<String>): String? {
        val candidates = when (field) {
            CsvImportableField.ProductId -> return guessProductIdColumn(headers)
            CsvImportableField.Barcode -> listOf("barcode", "upc", "ean", "sku")
            CsvImportableField.Name -> listOf("name", "product name", "description")
            CsvImportableField.ProductType -> listOf("type", "product type", "category")
            CsvImportableField.Manufacturer -> listOf("manufacturer", "mfr", "brand")
            CsvImportableField.Price -> listOf("price", "regular price", "sale price")
            CsvImportableField.Stock -> listOf("stock", "qty", "quantity", "on hand")
            CsvImportableField.PullCount -> listOf("pull count", "pull", "count")
        }
        return headers.firstOrNull { header ->
            val normalized = normalizeHeader(header)
            candidates.any { normalized == it }
        }
    }

    private fun nonEmptyLines(content: String): List<String> =
        content.lines().map { it.trim() }.filter { it.isNotEmpty() }

    private fun looksLikeHeader(fields: List<String>): Boolean {
        val normalized = fields.map(::normalizeHeader)
        if (normalized.any { it.contains("product") && it.contains("id") }) return true
        val keywords = listOf("name", "barcode", "type", "manufacturer", "price", "stock", "pull")
        return normalized.any { field -> keywords.any { field.contains(it) } }
    }

    private fun guessProductIdColumn(headers: List<String>): String? =
        headers.firstOrNull { header ->
            val normalized = normalizeHeader(header)
            normalized == "productid" ||
                normalized == "product id" ||
                normalized == "id" ||
                (normalized.contains("product") && normalized.contains("id"))
        }

    private fun normalizeHeader(header: String): String =
        header.lowercase().replace('_', ' ').replace('-', ' ').trim()

    private fun parseProductId(raw: String): Int? {
        val trimmed = raw.trim()
        trimmed.toIntOrNull()?.let { return it }
        val digits = trimmed.filter { it.isDigit() }
        return digits.toIntOrNull()
    }

    private fun stringValue(
        field: CsvImportableField,
        row: List<String>,
        parsed: ParsedCsv,
        mapping: CsvFieldMapping,
    ): String? {
        if (!mapping.enabledFields.contains(field)) return null
        val column = mapping.columnByField[field] ?: return null
        val index = parsed.headers.indexOf(column)
        if (index < 0 || index >= row.size) return null
        val value = row[index].trim()
        return value.ifEmpty { null }
    }

    private fun intValue(
        field: CsvImportableField,
        row: List<String>,
        parsed: ParsedCsv,
        mapping: CsvFieldMapping,
    ): Int? = stringValue(field, row, parsed, mapping)?.filter { it.isDigit() }?.toIntOrNull()
}
